### ISSUE TITLE
a balance update to zero after PSM trade never published

### ISSUE THREAD

**Issue by turadg** [Agoric/Agoric#6321]
## What is the Problem Being Solved?

In Ollinet @rowgraus made a trade and the "want" side of the transaction published a balance update (IST up) but give did not (aUSD down). 

https://gist.github.com/turadg/72f81dcf32a5b86cab74377f4a1090cd

A subsequent attempt to trade confirmed the aUSD funds were no longer available, so integrity is preserved, but the UI shows an incorrect state. According to RPC output, the wallet got IST for free, giving up no aUSD.

After making another trade (in reverse) the published balances were correct again.

## Description of the Design

Finish diagnosing the problem: where did the balance update drop?
1. Do the logs show what happened to the balance update message?
2. Did the contract call the publisher?
3. Did the RPC node miss it?
4. Is the code querying RPC missing it?

We've talked about publishing full states periodically (https://github.com/Agoric/agoric-sdk/issues/6184) but we've always assumed that every update will be published and queryable.

**Comment by turadg** [comment]
@arirubinstein finding the slogfile. @warner to help look through it to answer the questions above

**Comment by Chris-Hibbert** [comment]
[Ari reports](https://github.com/Agoric/agoric-sdk/issues/6184#issuecomment-1258403328) this is the slog file


https://drive.google.com/drive/folders/1H1OqSnASeI6OqoP0yHC2-d7kPBiREuUI?usp=sharing

**Comment by warner** [comment]
When the AUSD `VBANK_GRAB` message is sent through the bridge device to the cosmos side, the return value (with `VBANK_BALANCE_UPDATE`) doesn't list the AUSD balance:

```
{"type":"deliver","crankNum":5204,"vatID":"v1","deliveryNum":246,"replay":false,"kd":["message","ko123",{"methargs":{"body":"[\"toBridge\",[\"bank\",{\"amount\":\"149955\",\"denom\":\"ibc/usdc1234\",\"sender\":\"agoric1gwzya5h0gkzrqve5mjmqklkl3tnhh6nyqfwnhv\",\"type\":\"VBANK_GRAB\"}]]","slots":[]},"result":"kp1369"}],"vd":["message","o+12",{"methargs":{"body":"[\"toBridge\",[\"bank\",{\"amount\":\"149955\",\"denom\":\"ibc/usdc1234\",\"sender\":\"agoric1gwzya5h0gkzrqve5mjmqklkl3tnhh6nyqfwnhv\",\"type\":\"VBANK_GRAB\"}]]","slots":[]},"result":"p-139"}],"time":1663963985.306593,"monotime":85539.061685261}
{"type":"syscall","crankNum":5204,"vatID":"v1","deliveryNum":246,"syscallNum":0,"replay":false,"ksc":["invoke","kd30","callOutbound",{"body":"[\"bank\",{\"amount\":\"149955\",\"denom\":\"ibc/usdc1234\",\"sender\":\"agoric1gwzya5h0gkzrqve5mjmqklkl3tnhh6nyqfwnhv\",\"type\":\"VBANK_GRAB\"}]","slots":[]}],"vsc":["callNow","d-70","callOutbound",{"body":"[\"bank\",{\"amount\":\"149955\",\"denom\":\"ibc/usdc1234\",\"sender\":\"agoric1gwzya5h0gkzrqve5mjmqklkl3tnhh6nyqfwnhv\",\"type\":\"VBANK_GRAB\"}]","slots":[]}],"time":1663963985.309273,"monotime":85539.064366204}
{"type":"syscall-result","crankNum":5204,"vatID":"v1","deliveryNum":246,"syscallNum":0,"replay":false,"ksr":["ok",{"body":"{\"nonce\":48,\"type\":\"VBANK_BALANCE_UPDATE\",\"updated\":[]}","slots":[]}],"vsr":["ok",{"body":"{\"nonce\":48,\"type\":\"VBANK_BALANCE_UPDATE\",\"updated\":[]}","slots":[]}],"time":1663963985.310633,"monotime":85539.065726677}
{"type":"clist","crankNum":5204,"mode":"drop","vatID":"v1","kobj":"kp1369","vobj":"p-139","time":1663963985.312316,"monotime":85539.067408676}
{"type":"syscall","crankNum":5204,"vatID":"v1","deliveryNum":246,"syscallNum":1,"replay":false,"ksc":["resolve","v1",[["kp1369",false,{"body":"{\"nonce\":48,\"type\":\"VBANK_BALANCE_UPDATE\",\"updated\":[]}","slots":[]}]]],"vsc":["resolve",[["p-139",false,{"body":"{\"nonce\":48,\"type\":\"VBANK_BALANCE_UPDATE\",\"updated\":[]}","slots":[]}]]],"time":1663963985.31258,"monotime":85539.06767248799}
{"type":"syscall-result","crankNum":5204,"vatID":"v1","deliveryNum":246,"syscallNum":1,"replay":false,"ksr":["ok",null],"vsr":["ok",null],"time":1663963985.313223,"monotime":85539.068315285}
{"type":"deliver-result","crankNum":5204,"vatID":"v1","deliveryNum":246,"replay":false,"dr":["ok",null,{"meterType":"xs-meter-17","currentHeapCount":318273,"compute":73985,"allocate":50462752,"timestamps":[1663963985.306969,1663963985.308855,1663963985.310855,1663963985.311786,1663963985.313393,1663963985.313731]}],"time":1663963985.314291,"monotime":85539.06938337999}
```

I'm wondering if the cosmos side treats "balance just became zero" to mean "just delete the entire record" (because we certainly don't want to remember a bunch of empty accounts forever). But then a side-effect of that decision is that the VBANK_BALANCE_UPDATE comes back empty, instead of having a final `AUSD: 0` in it.

**Comment by Chris-Hibbert** [comment]
We were able to reproduce by attempting to withdraw max IST using the UI. This calculates fees so you'll spend all your AUSD.  We have a hypothesis that the go layer may be refusing to report zero balances.

**Comment by Chris-Hibbert** [comment]
@JimLarson confirmed that this particular request went through and operation that was reporting on all the balances in an account. Since there can be many balances, most of which are usually zero, it omits the zeros.  
In this particular case, we've just adjusted the balance of one coin, so if the balance for that coin is zero, it should be reported. In the more general case, javascript may need to poll for all purses that it knows about for a particular customer.

**Comment by warner** [comment]
With @JimLarson 's help we tracked this down to a working-as-designed-but-could-be-better behavior on the cosmos/go side. There is a Go function named [`getBalanceUpdate()`](https://github.com/Agoric/agoric-sdk/blob/b4ed09b92f03abdaeb8f4befd31b11e0d3586aa7/golang/cosmos/x/vbank/vbank.go#L73) which basically takes a list of addresses and produces a map that includes information about all non-zero balances owned by those addresses. This map does not mention the non-zero balances because we assume that there could be thousands of denoms floating around, and each user has never interacted with most of them, so we don't want to spend the space/time on a huge list of zeros.

This function is used in the [VBANK_GRAB](https://github.com/Agoric/agoric-sdk/blob/b4ed09b92f03abdaeb8f4befd31b11e0d3586aa7/golang/cosmos/x/vbank/vbank.go#L139) function that vat-vbank uses when it wants to `withdraw()` a `Payment` from one of those purses, which causes the cosmos-side balance to drop by some amount. (It is also used in the `VBANK_GIVE` used during `deposit()`, which makes the balance grow by some amount, but of course that cannot cause the balance to become zero). If you `VBANK_GRAB` *all* of the tokens, the balance drops to zero, and the return value (`VBANK_BALANCE_UPDATE`) omits that denom entirely, which means the vat-vbank Notifier for that address+denom pair is not updated, which means the chain-store node is not updated, which makes the wallet think that the AUSD balance is still present, when really it's been zeroed. This will happen any time you use an Agoric contract to completely drain a balance that lives in the cosmos bank/vbank modules.

This same function is used in the [BEGIN_BLOCK code](https://github.com/Agoric/agoric-sdk/blob/b4ed09b92f03abdaeb8f4befd31b11e0d3586aa7/golang/cosmos/x/vbank/module.go#L146-L155) that detects cosmos-side token transfers and notifies vat-vbank about the balances that were changed. It accumulates a list of touched addresses, then notifies JS about all the non-zero balances. This will suffer from the same problem: JS is not told about zeroes, so it won't update the Notifier, so the wallet will incorrectly keep reporting the last non-zero value. This could happen if you used the cosmos CLI (rather than an Agoric `offer`) to drain your account of BLD or some IBC-backed token by sending it to someone else. The wallet would not show your balance drop to zero.

But `getBalanceUpdate()` is *not* used in the [`VBANK_GET_BALANCE`](https://github.com/Agoric/agoric-sdk/blob/b4ed09b92f03abdaeb8f4befd31b11e0d3586aa7/golang/cosmos/x/vbank/vbank.go#L122) query, which asks for a single address/denom pair. That query *should* be able to return a zero.

@JimLarson 's planned fix is to change the way `getBalanceUpdate` is used so that it accepts a list of `(addr, denom)` pairs (or maybe a map from `addr` to ` list of denoms), and reports all the indicated balances, including the zeros.

**Comment by JimLarson** [comment]
Tentatitive solution: on the Cosmos side we can track which denoms might have been affected by an operation, then report possibly-zero balances for all and only those denoms.

Issue 1: @michaelfig do you know if we can reliably read the `amount` field in the `transfer` events that we monitor?

Issue 2: @warner @Chris-Hibbert @dckc We currently report balances for all nonzero denoms in the account, regardless of whether they might have been changed. Is it okay if we omit the known-unchanged nonzero balances?

**Comment by dckc** [comment]
> Is it okay if we omit the known-unchanged nonzero balances?

yes, I think so.

**Comment by JimLarson** [comment]
Fixed in #6351.
